File: leetcode/637.cs
using System;
using System.Collections.Generic;
namespace leetcode637
{
    public class TreeNode
    {
        public int val;
        public TreeNode left, right;
        public TreeNode(int x) { val = x; }
    }
    public class Solution
    {
        public IList<double> AverageOfLevels(TreeNode root)
        {
            return level(root);
        }

        public IList<double> level(TreeNode root)
        {
            Queue<int> q1 = new Queue<int>();
            Queue<TreeNode> q2 = new Queue<TreeNode>();
            Dictionary<int, Int64> d1 = new Dictionary<int, Int64>();
            Dictionary<int, int> d2 = new Dictionary<int, int>();
            q1.Enqueue(1);
            q2.Enqueue(root);
            while (q1.Count != 0 && q2.Count != 0)
            {
                TreeNode node = q2.Dequeue();
                int l = q1.Dequeue();
                if (d1.ContainsKey(l))
                {
                    d1[l] += node.val;
                }
                else
                {
                    d1.Add(l, node.val);
                }
                if (d2.ContainsKey(l))
                {
                    d2[l]++;
                }
                else
                {
                    d2.Add(l, 1);
                }
                if (node.left != null)
                {
                    q1.Enqueue(l + 1);
                    q2.Enqueue(node.left);
                }
                if (node.right != null)
                {
                    q1.Enqueue(l + 1);
                    q2.Enqueue(node.right);
                }
            }
            List<double> list = new List<double>();
            foreach (KeyValuePair<int, Int64> kv in d1)
            {
                list.Add((double)(kv.Value) / (d2[kv.Key]));
            }
            return list;
        }
    }
}

File: leetcode/637.cpp
#include "leetcode.h"
#include <queue>
#include <unordered_map>
class Solution
{
public:
    vector<double> averageOfLevels(TreeNode *root)
    {
        return levelorder(root);
        typedef pair<TreeNode *, int> node;
        unordered_map<int, double> sumrecord;
        unordered_map<int, int> numrecord;
        queue<node> q;
        q.push(node(root, 0));
        while (!q.empty())
        {
            auto n = q.front();
            q.pop();
            sumrecord[n.second] += n.first->val;
            numrecord[n.second]++;
            if (n.first->left != nullptr)
                q.push(node(n.first->left, n.second + 1));
            if (n.first->right != nullptr)
                q.push(node(n.first->right, n.second + 1));
        }
        vector<double> v(numrecord.size());
        for (auto &x : sumrecord)
        {
            v[x.first] = x.second / numrecord[x.first];
        }
        return v;
    }

    vector<double> levelorder(TreeNode *root)
    {
        vector<double> v;
        queue<TreeNode *> q;
        q.push(root);

        while (!q.empty())
        {
            queue<TreeNode *> nextlevel;
            int64_t sum = 0;
            int counter = 0;
            while (!q.empty())
            {
                auto p = q.front();
                q.pop();
                sum += p->val, counter++;
                if (p->left != nullptr)
                    nextlevel.push(p->left);
                if (p->right != nullptr)
                    nextlevel.push(p->right);
            }
            q = nextlevel;
            v.push_back(sum * 1.0 / counter);
        }
        return v;
    }
};
